azure-iot-device/tests/common/pipeline/pipeline_stage_test.py

```diff
@@ -167,6 +187,81 @@ def test_passes_event_to_previous_stage_which_throws_base_exception(
                 stage.handle_pipeline_event(event)
             assert unhandled_error_handler.call_count == 0
 
-            pass
-
     setattr(module, "Test{}UnknownEvents".format(cls.__name__), LocalTestObject)
+
+
+class ThreadLaunchedError(Exception):
+    pass
+
+
+def add_pipeline_thread_tests(
+    cls, module, methods_that_enter_pipeline_thread, methods_that_can_run_in_any_thread
+):
+    def does_method_assert_pipeline_thread(method_name):
+        if method_name.startswith("__"):
+            return False
+        elif method_name in methods_that_enter_pipeline_thread:
+            return False
+        elif method_name in methods_that_can_run_in_any_thread:
+            return False
+        else:
+            return True
+
+    methods_that_assert_pipeline_thread = [
+        x[0]
+        for x in inspect.getmembers(cls, inspect.isfunction)
+        if does_method_assert_pipeline_thread(x[0])
+    ]
+
+    @pytest.mark.describe("{} - Pipeline threading".format(cls.__name__))
+    class LocalTestObject(object):
+        @pytest.fixture
+        def stage(self):
+            return cls()
+
+        @pytest.mark.parametrize("method_name", methods_that_assert_pipeline_thread)
+        @pytest.mark.it("Has a method that must be in a pipeline thread")
```

Comment:
This is not a great requirement name, but I do see your dilemma - you test pipeline objects as objects, but the thread-running-ness of its methods is kind of within the domain of a method unit. So instead you have to parametrize the pipeline object methods and show that each is called in the correct thread.

Since obviously we don't want to revisit the entire testing approach, maybe a slightly better name? I might suggest `Enforces use of the pipeline thread when calling method`?